service/matching/matchingEngine.go

```diff
@@ -422,50 +426,94 @@ func (e *matchingEngineImpl) QueryWorkflow(ctx context.Context, queryRequest *m.
 	taskListName := queryRequest.TaskList.GetName()
 	taskList := newTaskListID(domainID, taskListName, persistence.TaskListTypeDecision)
 	taskListKind := common.TaskListKindPtr(queryRequest.TaskList.GetKind())
-	tlMgr, err := e.getTaskListManager(taskList, taskListKind)
-	if err != nil {
-		return nil, err
-	}
-	queryTask := &queryTaskInfo{
-		queryRequest: queryRequest,
-		taskID:       uuid.New(),
-	}
-	err = tlMgr.SyncMatchQueryTask(ctx, queryTask)
-	if err != nil {
-		return nil, err
-	}
 
-	queryResultCh := make(chan *workflow.RespondQueryTaskCompletedRequest, 1)
-	e.queryMapLock.Lock()
-	e.queryTaskMap[queryTask.taskID] = queryResultCh
-	e.queryMapLock.Unlock()
-	defer func() {
+query_loop:
+	for {
+		tlMgr, err := e.getTaskListManager(taskList, taskListKind)
+		if err != nil {
+			return nil, err
+		}
+		queryTask := &queryTaskInfo{
+			queryRequest: queryRequest,
+			taskID:       uuid.New(),
+		}
+		err = tlMgr.SyncMatchQueryTask(ctx, queryTask)
+		if err != nil {
+			return nil, err
+		}
+
+		queryResultCh := make(chan *queryResult, 1)
 		e.queryMapLock.Lock()
-		delete(e.queryTaskMap, queryTask.taskID)
+		e.queryTaskMap[queryTask.taskID] = queryResultCh
 		e.queryMapLock.Unlock()
-	}()
+		defer func() {
+			e.queryMapLock.Lock()
+			delete(e.queryTaskMap, queryTask.taskID)
+			e.queryMapLock.Unlock()
+		}()
+
+		select {
+		case result := <-queryResultCh:
+			if result.err == nil {
+				return &workflow.QueryWorkflowResponse{QueryResult: result.result}, nil
+			}
+
+			if result.err == errQueryBeforeFirstDecisionCompleted {
+				// query before first decision completed, so wait for first decision task to complete
+				expectedNextEventID := result.waitNextEventID
```

Comment:
Can we initial expectedNextEventID to some value like 'firstEventID' or something before we call GetMutableState first time.  Then we don't need to return it from QueryResult.  Then once wait_loop calls GetMutableState we set it for future calls.

Reply:
That would require an additional round trip from matching to history. We do GetMutableState() to block and wait for the "expectedNextEventID". If we initialize it to firstEventID, the first call to GetMutableState() will not block and is a waste.